### DISCUSSION TITLE
Token is returned if login is invoked automatically

### DISCUSSION THREAD

**Discussion by madtag** [soofstad/react-oauth2-pkce#88] [Q&A] [answered]
Hi all,

I am trying to use a 'useEffect' to invoke the login process instead of a button click as in the example.  It works fine if the user is not previously authenticated.  But when the user is logged in, and hits logout the token goes to empty string as expected, but when the login function is invoked, the user is not redirected to the keycloak login page and the token is returned.

```tsx
React.useEffect(() => {
  if (token == null || token.length === 0) {
    console.log("Attempting to login")
    console.log("Login => ", login())
  }
}, [login, token])
```

In the image below you see what happens after logout is invoked. 

![image](https://user-images.githubusercontent.com/5601734/235859889-2dc1b798-afc4-4141-aad9-3ead0470dcaf.png)

Is this an issue with keycloak not logging out fast enough? Is there some suggestion on what I could do to solve this?

**Comment by sebastianvitterso** [comment]
On a more general level, what are you trying to achieve? 

If the token changes, and is set to `null` or `""`, that probably means you have been logged out. Does this mean that you want to automatically log in (via redirect to keycloak) whenever the user logs out?

If you post your use case, it might be easier for us to understand what you're trying to do, and we can aid in finding a solution. I don't see any immediate issues in the snippet you posted, but there might be other issues.

**Reply by madtag** [reply]
Yes exactly. I am invoking logout (through a button), token becomes empty string and I invoke login.  But the token is getting updated without being redirected to the login screen.

The usecase is to be redirected to the login screen automatically and not have to press a button (or link) manually when you are not logged in or when you press logout.

**Comment by soofstad** [comment, marked as answer]
This will also trigger a new login whenever the token is refreshed, which happens in the background continuously. You probably don't want to redirect your users every 1-10min.

Have you looked into the "autoLogin" config parameter? Any reason you can't use this?

If you need some more help, posting the "authConfig" and some more context would be useful :slightly_smiling_face:

**Reply by madtag** [reply]
So, my usecase is that I don't want the user to have to press a button to be redirected to the login screen.  If they are not authenticated they should be redirected automatically.  I have autologin set to false.  In my code if the token changes it doesn't invoke login unless the token becomes an empty string or null.

**Reply by madtag** [reply]
All i needed to do was set autologin to true and I don't need to use that useEffect! Thanks a million. Problem solved
